pkg/kubectl/cmd/portforward.go

```diff
@@ -118,20 +125,43 @@ func (o *PortForwardOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, arg
 	var err error
 	o.PodName = cmdutil.GetFlagString(cmd, "pod")
 	if len(o.PodName) == 0 && len(args) == 0 {
-		return cmdutil.UsageErrorf(cmd, "POD is required for port-forward")
+		return cmdutil.UsageErrorf(cmd, "POD or RESOURCE/NAME is required for port-forward")
+	}
+
+	o.Namespace, _, err = f.DefaultNamespace()
+	if err != nil {
+		return err
 	}
 
+	builder := f.NewBuilder().
+		Internal().
+		ContinueOnError().
+		NamespaceParam(o.Namespace).DefaultNamespace()
+
 	if len(o.PodName) != 0 {
 		printDeprecationWarning("port-forward POD", "-p POD")
 		o.Ports = args
 	} else {
-		o.PodName = args[0]
-		o.Ports = args[1:]
-	}
+		getPodTimeout, err := cmdutil.GetPodRunningTimeoutFlag(cmd)
+		if err != nil {
+			return cmdutil.UsageErrorf(cmd, err.Error())
+		}
 
-	o.Namespace, _, err = f.DefaultNamespace()
-	if err != nil {
-		return err
+		resourceName := args[0]
+		builder.ResourceNames("pods", resourceName)
+
+		obj, err := builder.Do().Object()
+		if err != nil {
+			return err
+		}
+
+		forwardablePod, err := f.AttachablePodForObject(obj, getPodTimeout)
```

Comment:
This resolves to a Pod at the time that port-forward is started.  This may be a little confusing to users as it (a) won't load balance across all the pods in a service and (b) won't re-attach to a good pod if this pod dies.  That is especially useful across updates of a Deployment.

Regardless -- this is better than the current situation but it should probably be documented as a limitation in the command help.

Reply:
Regarding re-attach to a good pod, I see port-forward as a pod operation, and the use of resource/name to select a pod is just for convenience.  I can create a separate issue to track that if that is desirable.

Reply:
@jbeda PTAL